// metasync/dist/lib/do.js
'use strict';

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function Do() {}

var chain = function chain(fn) {
  var current = function current(done) {
    if (done) current.done = done;

    if (current.prev) {
      current.prev.next = current;
      current.prev();
    } else {
      current.forward();
    }

    return current;
  };

  var prev = this instanceof Do ? this : null;

  for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    args[_key - 1] = arguments[_key];
  }

  var fields = {
    prev: prev,
    fn: fn,
    args: args,
    done: null
  };
  Object.setPrototypeOf(current, Do.prototype);
  return Object.assign(current, fields);
};

Do.prototype["do"] = function (fn) {
  for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
    args[_key2 - 1] = arguments[_key2];
  }

  return chain.call.apply(chain, [this, fn].concat(args));
};

Do.prototype.forward = function () {
  var _this = this;

  if (this.fn) {
    this.fn.apply(this, _toConsumableArray(this.args).concat([function (err, data) {
      var next = _this.next;

      if (next) {
        if (next.fn) next.forward();
      } else if (_this.done) {
        _this.done(err, data);
      }
    }]));
  }
};

module.exports = {
  "do": chain
};
