// metasync/dist/lib/poolify.opt.js
'use strict';

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var duplicate = function duplicate(factory, n) {
  return Array.from({
    length: n
  }, factory);
};

var provide = function provide(callback) {
  return function (item) {
    setImmediate(function () {
      callback(item);
    });
  };
};

var poolify = function poolify(factory, min, norm, max) {
  var allocated = norm;
  var items = duplicate(factory, norm);
  var delayed = [];

  var pool = function pool(par) {
    if (Array.isArray(par)) {
      while (par.length) {
        var item = par.shift();
        var request = delayed.shift();
        if (request) request(item);else items.push(item);
      }

      return pool;
    }

    if (items.length < min && allocated < max) {
      var grow = Math.min(max - allocated, norm - items.length);
      allocated += grow;
      var instances = duplicate(factory, grow);
      items.push.apply(items, _toConsumableArray(instances));
    }

    var res = items.pop();
    if (!par) return res;
    var callback = provide(par);
    if (res) callback(res);else delayed.push(callback);
    return pool;
  };

  return pool;
};

module.exports = {
  poolify: poolify
};
